st2common/st2common/garbage_collection/inquiries.py

```diff
@@ -0,0 +1,94 @@
+from st2common.constants import action as action_constants
+from st2common.models.db.auth import UserDB
+from st2common.persistence.execution import ActionExecution
+from st2common.services import action as action_service
+from st2common.services import executions
+from st2common.util import action_db as action_utils
+from st2common.util.action_db import (get_action_by_ref, get_runnertype_by_name)
+from st2common.util.date import get_datetime_utc_now
+
+__all__ = [
+    'purge_inquiries',
+]
+
+
+def purge_inquiries(logger):
+    """Purge Inquiries that have exceeded their configured TTL
+
+    At the moment, Inquiries do not have their own database model, so this function effectively
+    is another, more specialized GC for executions. It will look for executions with a 'pending'
+    status that use the 'inquirer' runner, which is the current definition for an Inquiry. Then
+    it will force-fail those that:
+
+    - Have a nonzero TTL
+    - Have existed longer than their TTL
+    """
+
+    # Get all existing Inquiries
+    filters = {'runner__name': 'inquirer', 'status': action_constants.LIVEACTION_STATUS_PENDING}
+    inquiries = list(ActionExecution.query(**filters))
+
+    gc_count = 0
+
+    # Inspect each Inquiry, and determine if TTL is expired
+    for inquiry in inquiries:
+
+        ttl = int(inquiry.result.get('ttl'))
+        min_since_creation = int(
+            (get_datetime_utc_now() - inquiry.start_timestamp).total_seconds() / 60
+        )
+
+        if ttl <= 0:
+            logger.debug("Inquiry %s has a TTL of %s. Skipping." % (inquiry.id, ttl))
+            continue
+
+        logger.debug("Inquiry %s has a TTL of %s and was started %s minute(s) ago" % (
+                     inquiry.id, ttl, min_since_creation))
+
+        if min_since_creation > ttl:
+
+            gc_count += 1
+            logger.info("TTL expired for Inquiry %s. Marking as timed out." % inquiry.id)
+
+            liveaction_db = action_utils.update_liveaction_status(
+                status=action_constants.LIVEACTION_STATUS_TIMED_OUT,
+                result=inquiry.result,
+                liveaction_id=inquiry.liveaction.get('id'))
+            executions.update_execution(liveaction_db)
+
+            # Call Inquiry runner's post_run to trigger callback to workflow
+            runner_container = get_runner_container()
+            action_db = get_action_by_ref(liveaction_db.action)
+            runnertype_db = get_runnertype_by_name(action_db.runner_type['name'])
+            runner = runner_container._get_runner(runnertype_db, action_db, liveaction_db)
+            runner.post_run(status=action_constants.LIVEACTION_STATUS_TIMED_OUT,
+                            result=inquiry.result)
+
+            if liveaction_db.context.get("parent"):
+
+                # Request that root workflow resumes
+                root_liveaction = action_service.get_root_liveaction(liveaction_db)
+                action_service.request_resume(
```

Comment:
I suggest parent workflow cancellation here.

Reply:
Hrm, can you elaborate? Not sure I agree - allowing the workflow to resume allows the user to use `on-success` or `on-failure` as needed to deal with folks not responding on time.

Reply:
I suggested cancel because it's not technically an error.  But you're right. I can see clean up happening on-failure.  Let's keep this as is.